File: web/src/lib/prompts.ts
/**
 * Prompts for the HireFlow AI agent (interview-time).
 *
 * These mirror the screening prompts in `api/hireflow/prompts.py` so the agent
 * judges interview evidence by exactly the same policy the resume screening
 * used. Two rules run through all of them:
 *
 *   1. The model proposes evidence; it never adjudicates it. Quotes must be
 *      copied verbatim so a program can check them.
 *   2. The four-state status policy is spelled out, because left to its own
 *      judgement the model collapses "related evidence exists but doesn't prove
 *      it" into "absent" — which loses the distinction the product sells.
 */

export const STATUS_POLICY = `Status must be exactly one of these four values, chosen by these rules:

  "met"        Explicit evidence satisfies the whole requirement.
  "partial"    Explicit evidence satisfies PART of the requirement but not all of it.
  "unverified" Related or adjacent evidence exists, but it does not demonstrate
               the specific thing the requirement asks for.
  "absent"     No relevant evidence of any kind.

If ANY adjacent or partially relevant evidence exists, prefer "unverified" over
"absent". Reserve "absent" for genuine silence on the topic.

RULES THAT OVERRIDE NAIVE KEYWORD MATCHING:
  - Naming a technology is a CLAIM, not evidence. Demonstrated work beats a keyword.
  - "In production" means production. A university or personal project is at most
    "partial", usually "unverified", for a requirement asking for production use.
  - Judge the requirement AS STATED, not the candidate's overall strength.
    Adjacent strength in another stack does not satisfy a requirement they lack.`;

/**
 * Re-evaluate one requirement against a recorded interview answer.
 * This is the call that can promote a requirement to "met".
 */
export function interviewPrompt(input: {
  candidateName: string;
  jobTitle: string;
  requirementId: string;
  requirementText: string;
  requirementKind: string;
  priorStatus: string;
  priorReason: string;
  missingDetail?: string | null;
  question: string;
  answer: string;
}): string {
  return `You are re-evaluating ONE job requirement for a candidate after an interview answer.

CANDIDATE: ${input.candidateName}
ROLE: ${input.jobTitle}

REQUIREMENT ${input.requirementId} (${input.requirementKind}): ${input.requirementText}

STATUS FROM RESUME: ${input.priorStatus}
WHY (from resume screening): ${input.priorReason}
WHAT WAS MISSING: ${input.missingDetail ?? "not specified"}

INTERVIEW QUESTION THAT WAS ASKED:
${input.question}

CANDIDATE'S RECORDED ANSWER:
---
${input.answer}
---

Re-evaluate the requirement using the recorded answer as new evidence.

${STATUS_POLICY}

Rules specific to interview evidence:
- The answer is recorded testimony. Treat CONCRETE, SPECIFIC claims as strong
  evidence: named tools, actions the candidate personally performed, systems they
  operated, incidents they resolved, scale, outcomes. If the answer establishes
  the requirement with that kind of specificity, new_status becomes "met".
- Treat a VAGUE confirmation as insufficient. "Yes, I've used it", "I'm familiar
  with it", "a bit" does NOT establish the requirement. Keep the prior status and
  you MUST return a follow_up asking for the specific missing detail.
- If the answer is off-topic or does not address the requirement, keep the prior
  status and return a follow_up.
- evidence_quote must be copied CHARACTER-FOR-CHARACTER from the CANDIDATE'S
  RECORDED ANSWER above — not from the resume, not paraphrased, at least 4 words.
  A program verifies it against the answer text and will reject a quote it cannot
  locate. If nothing suitable exists, use null.

Return ONLY this JSON:
{"new_status": "met|partial|unverified|absent", "sufficient": true,
 "reason": "one or two sentences for the recruiter explaining the decision",
 "evidence_quote": "verbatim span from the answer, or null",
 "missing_detail": "what is still missing, or null",
 "follow_up": "the next question to ask, or null"}`;
}

/** Generate a targeted validation question for one open requirement. */
export function questionPrompt(input: {
  candidateName: string;
  jobTitle: string;
  requirementId: string;
  requirementText: string;
  priorStatus: string;
  priorReason: string;
  missingDetail?: string | null;
  alreadyAsked: string[];
}): string {
  const asked = input.alreadyAsked.length
    ? input.alreadyAsked.map((q) => `- ${q}`).join("\n")
    : "- (none yet)";

  return `You are writing ONE interview question to close a specific evidence gap.
This is not a generic role question — it exists to resolve a requirement the
resume failed to establish.

CANDIDATE: ${input.candidateName}
ROLE: ${input.jobTitle}

REQUIREMENT ${input.requirementId}: ${input.requirementText}
CURRENT STATUS: ${input.priorStatus}
WHY: ${input.priorReason}
WHAT IS MISSING: ${input.missingDetail ?? "not specified"}

QUESTIONS ALREADY ASKED (do not repeat or closely paraphrase these):
${asked}

Write a question that:
  - references what the resume DID show, where relevant, so the candidate can
    respond concretely rather than feeling tested
  - asks directly about the missing detail
  - is open enough that a strong candidate can demonstrate depth
  - does not lead the candidate toward the answer you want

Also write follow_up: the question to ask if the first answer still leaves the
requirement unproven. Use null only if no sensible follow-up exists.
rationale: one sentence for the recruiter explaining why this question exists.

Return ONLY this JSON:
{"question": "...", "rationale": "...", "follow_up": "..."}`;
}

/**
 * Recruiter Q&A about the active candidate.
 *
 * The recruiter can ask anything; these are the intents the UI offers as
 * shortcuts. The agent gets the full screening + interview state for ONE
 * candidate so its answer is grounded and cannot drift to another candidate.
 */
export const ASK_INTENTS = {
  why: "Explain why the focused requirement currently has the status it has.",
  missing:
    "State precisely what evidence is missing to establish the focused requirement.",
  explain:
    "Explain what the focused requirement is asking for and why it matters for this role.",
  followup:
    "Give the single best next question to ask about the focused requirement.",
  summary:
    "Summarise this candidate against the role: what is evidenced, what is not.",
  progress:
    "Review interview progress: what is now verified, what is still open, and what to do next.",
  next: "Recommend the single most useful next question to ask this candidate, considering everything already asked and answered.",
  free: "Answer the recruiter's question.",
} as const;

export type AskIntent = keyof typeof ASK_INTENTS;

export function askPrompt(input: {
  intent: AskIntent;
  userQuestion?: string;
  candidateName: string;
  candidateHeadline: string;
  candidateYears: number;
  candidateSummary: string;
  jobTitle: string;
  focusRequirement?: {
    id: string;
    text: string;
    kind: string;
    status: string;
    reason: string;
    missingDetail?: string | null;
  };
  requirementLines: string[];
  askedQuestions: string[];
  interviewLog: string[];
}): string {
  const focus = input.focusRequirement
    ? `FOCUSED REQUIREMENT ${input.focusRequirement.id} (${input.focusRequirement.kind}): ${input.focusRequirement.text}
  current status: ${input.focusRequirement.status}
  reason: ${input.focusRequirement.reason}
  missing: ${input.focusRequirement.missingDetail ?? "not specified"}`
    : "FOCUSED REQUIREMENT: none selected";

  const task =
    input.intent === "free" && input.userQuestion
      ? `THE RECRUITER ASKS: "${input.userQuestion}"`
      : `TASK: ${ASK_INTENTS[input.intent]}`;

  return `You are HireFlow AI, an interview assistant helping a recruiter evaluate ONE
candidate. Answer only about the candidate below, using only the state given.

ROLE: ${input.jobTitle}

CANDIDATE: ${input.candidateName} — ${input.candidateHeadline} (${input.candidateYears} years)
SCREENING SUMMARY: ${input.candidateSummary}

REQUIREMENT STATE (id | kind | status | requirement):
${input.requirementLines.map((l) => `  ${l}`).join("\n")}

${focus}

QUESTIONS ALREADY ASKED:
${input.askedQuestions.length ? input.askedQuestions.map((q) => `  - ${q}`).join("\n") : "  - (none yet)"}

INTERVIEW ACTIVITY SO FAR:
${input.interviewLog.length ? input.interviewLog.map((l) => `  - ${l}`).join("\n") : "  - (no answers analysed yet)"}

${task}

Rules:
- Be specific and grounded in the state above. Never invent evidence, employers,
  projects or numbers that are not present.
- Refer to requirements by their id (e.g. REQ-07) where relevant.
- Do not recommend a hiring decision — that belongs to the recruiter.
- Never change a requirement's status here. This is an explanation, not a verdict.
- 1–4 sentences. Plain, direct, no flattery, no bullet lists.

Return ONLY this JSON:
{"title": "a short label, at most 6 words", "body": "your answer"}`;
}

File: api/hireflow/prompts.py
"""
Prompts.

Two design rules run through all of these:

1. The model proposes evidence; it never adjudicates it. Every prompt asks for a
   quote *copied verbatim* from the source, because a copied quote is checkable
   and a paraphrase is not.
2. The four-state status policy is spelled out as rules. Left to its own
   judgement the model collapses "related evidence exists but doesn't prove it"
   into "absent", which loses precisely the distinction the product sells.
"""

STATUS_POLICY = """\
Status must be exactly one of these four values, chosen by these rules:

  "met"        Explicit text in the resume satisfies the whole requirement.
  "partial"    Explicit text satisfies PART of the requirement but not all of it.
  "unverified" Related or adjacent evidence exists, but it does not demonstrate
               the specific thing the requirement asks for. Use this when the
               resume gets close without proving it. Example: the requirement is
               production Kubernetes and the resume only shows Docker — Docker is
               related, so this is "unverified", NOT "absent".
  "absent"     No relevant evidence of any kind appears in the resume.

The distinction between "unverified" and "absent" matters and is often wrong.
If ANY adjacent or partially relevant text exists, prefer "unverified".
Reserve "absent" for genuine silence on the topic.

THREE RULES THAT OVERRIDE NAIVE KEYWORD MATCHING:

1. A SKILLS LIST IS A CLAIM, NOT EVIDENCE.
   If a technology appears only in a "Skills" or "Technologies" list and is NOT
   demonstrated anywhere in the Experience or Projects sections, the status is
   "unverified", never "met". Say so in the reason, e.g. "Kubernetes appears in
   the skills list but no experience or project demonstrates it." Real work
   beats a keyword every time.

2. "IN PRODUCTION" MEANS PRODUCTION.
   When a requirement asks for production experience, evidence from a university
   project, a personal project, a course, or local-only usage is at most
   "partial", and usually "unverified". A capstone project that "uses Spring
   Boot" does not satisfy "production experience with Spring Boot".

3. JUDGE THE REQUIREMENT AS STATED, NOT OVERALL STRENGTH.
   A candidate who is excellent in a different stack does not satisfy a
   requirement they lack. Years in one language do NOT count toward a
   requirement for a different language. If the role requires Java and the
   candidate has one year of Java on a legacy service but seven years of Python,
   the Java requirement is "partial" (thin, real) or "absent" — not "met" —
   regardless of how strong they are overall. Do not reward adjacent strength."""

EVIDENCE_RULE = """\
evidence_quote must be a span of text COPIED CHARACTER-FOR-CHARACTER from the
RESUME above. Do not paraphrase, summarise, reflow or correct it. Do not invent
a quote. It must be at least 4 words long.
If no suitable span exists, set evidence_quote to null.
Prefer a quote from the Experience or Projects sections over one from a Skills
list. If the only mention of the skill is in a Skills list, you may quote that
list, but the status must then be "unverified" per rule 1 above, and
location_hint must say "Skills list".
A separate program will check your quote against the source document and will
reject your finding if the quote cannot be located, so copying exactly is in
your interest."""


REQUIREMENTS_PROMPT = """\
You are decomposing a job description into individually addressable requirements.

JOB DESCRIPTION:
---
{jd_text}
---

Extract requirements ONLY from the explicit requirement lists — sections titled
Requirements, Must have, Preferred, Nice to have, Qualifications, or similar
bulleted lists. IGNORE the mission statement, company/team description, and the
narrative "you will…" responsibilities in the intro paragraph. A sentence like
"You will design APIs and operate what you ship" is describing the job, not a
screenable requirement, so do not extract it.

Take one requirement per bullet, keeping the job's own grouping — if a single
bullet says "schema design and query optimisation", that is ONE requirement.
Do not atomise and do not invent requirements the lists do not contain.

For each requirement return:
  text         A single clear requirement, phrased as a capability.
  kind         "hard" if required, "soft" if listed as nice to have / preferred.
  source_quote A span copied verbatim from the JOB DESCRIPTION that this
               requirement came from.

Return ONLY this JSON:
{{"requirements": [{{"text": "...", "kind": "hard", "source_quote": "..."}}]}}"""


REVIEW_PROMPT = """\
You are screening one candidate against a fixed list of job requirements.

RESUME ({source_doc}):
---
{resume_text}
---

REQUIREMENTS:
{requirements_block}

For EVERY requirement id listed above, produce one finding. Do not skip any.

{status_policy}

{evidence_rule}

Also return, about the candidate overall:
  headline          Their current role and domain, e.g. "Senior Backend Engineer · Fintech"
  location          As stated in the resume
  years_experience  Integer. Total years of professional experience evidenced.
  summary           2-3 sentences for a recruiter. State plainly what is evidenced
                    and what is not. Do not flatter. Do not recommend a decision.

For each finding return:
  requirement_id   Exactly as given above (e.g. "REQ-03")
  status           Per the policy above
  reason           One or two sentences explaining the verdict, referring to what
                   the resume does and does not show.
  missing_detail   If status is not "met": precisely what is missing. Otherwise null.
  confidence       0.0 to 1.0
  evidence_quote   Verbatim span from the RESUME, or null
  location_hint    Where in the resume it came from, e.g. "Experience — Kaleidofin"

Return ONLY this JSON:
{{"headline": "...", "location": "...", "years_experience": 0, "summary": "...",
  "findings": [{{"requirement_id": "REQ-01", "status": "met", "reason": "...",
  "missing_detail": null, "confidence": 0.9, "evidence_quote": "...",
  "location_hint": "..."}}]}}"""


INTERVIEW_PROMPT = """\
You are re-evaluating ONE requirement for a candidate after an interview.

REQUIREMENT: {requirement_text}

STATUS FROM RESUME: {prior_status}
WHY (from resume): {prior_reason}

INTERVIEW QUESTION THAT WAS ASKED:
{question}

CANDIDATE'S RECORDED ANSWER:
---
{answer}
---

Re-evaluate the requirement using the recorded answer as new evidence.

{status_policy}

Rules specific to interview evidence:
- The answer is recorded testimony from the interview. Treat CONCRETE, SPECIFIC
  claims as strong evidence: named tools, actions the candidate personally
  performed, systems they operated, incidents they resolved, outcomes. If the
  answer establishes the requirement with that kind of specificity, status
  becomes "met".
- Treat a VAGUE confirmation as insufficient. "Yes, I've used it" or "I'm
  familiar with it" does not establish the requirement — that stays "unverified"
  and you MUST provide a follow_up question that asks for the specific missing
  detail.
- evidence_quote must be copied CHARACTER-FOR-CHARACTER from the CANDIDATE'S
  RECORDED ANSWER above (not from the resume). A program verifies it.

Return ONLY this JSON:
{{"new_status": "met", "reason": "...", "evidence_quote": "...",
  "missing_detail": null, "follow_up": null}}"""


QUESTIONS_PROMPT = """\
You are writing interview questions to close specific evidence gaps for one
candidate. These are not generic role questions — each one exists to resolve a
requirement the resume failed to establish.

CANDIDATE: {name}

OPEN REQUIREMENTS:
{gaps_block}

For each open requirement write a question that:
  - references the specific thing the resume DID show, where relevant, so the
    candidate can respond concretely rather than being tested
  - asks directly about the missing detail
  - is open enough that a strong candidate can demonstrate depth
  - does not lead the candidate toward the answer you want

Also write follow_up: the question to ask if the first answer still leaves the
requirement unproven. Set it to null only if no sensible follow-up exists.

rationale: one sentence, for the recruiter, explaining why this question exists.

Return ONLY this JSON:
{{"questions": [{{"requirement_id": "REQ-03", "question": "...",
  "rationale": "...", "follow_up": "..."}}]}}"""
